core/src/main/java/org/apache/brooklyn/enricher/stock/AbstractTransformer.java

```diff
@@ -59,27 +72,55 @@ public void setEntity(EntityLocal entity) {
         super.setEntity(entity);
 
         this.producer = getConfig(PRODUCER) == null ? entity: getConfig(PRODUCER);
-        this.sourceSensor = (Sensor<T>) getRequiredConfig(SOURCE_SENSOR);
+        this.sourceSensor = (Sensor<T>) getConfig(SOURCE_SENSOR);
         Sensor<?> targetSensorSpecified = getConfig(TARGET_SENSOR);
+        List<? extends Sensor<?>> triggerSensorsSpecified = getConfig(TRIGGER_SENSORS);
+        List<? extends Sensor<?>> triggerSensors = triggerSensorsSpecified != null ? triggerSensorsSpecified : ImmutableList.<Sensor<?>>of();
         this.targetSensor = targetSensorSpecified!=null ? (Sensor<U>) targetSensorSpecified : (Sensor<U>) this.sourceSensor;
-        if (producer.equals(entity) && targetSensorSpecified==null) {
+        if (targetSensor == null) {
+            throw new IllegalArgumentException("Enricher "+JavaClassNames.simpleClassName(this)+" has no "+TARGET_SENSOR.getName()+", and it cannot be inferred as "+SOURCE_SENSOR.getName()+" is also not set");
+        }
+        if (sourceSensor == null && triggerSensors.isEmpty()) {
+            throw new IllegalArgumentException("Enricher "+JavaClassNames.simpleClassName(this)+" has no "+SOURCE_SENSOR.getName()+" and no "+TRIGGER_SENSORS.getName());
+        }
```

Comment:
You could use triggerSensors in combination with sourceSensor, as an additional way of it recalculating a value from the sourceSensor. Not quite sure when you'd want to do that but it should work!